subsys/net/lib/download_client/src/download_client.c

```diff
@@ -9,1054 +9,550 @@
 
-	client->config = *config;
-	client->host = host;
-	client->close_when_done = false;
-	k_mutex_unlock(&client->mutex);
 	return 0;
 }
 
-int download_client_disconnect(struct download_client *const client)
+int download_client_start(struct download_client *dlc,
+			  const struct download_client_host_cfg *host_config,
+			  const char *uri, size_t from)
 {
-	if (client == NULL || is_idle(client)) {
+	int err;
+	struct dlc_transport *transport_connected = NULL;
+
+	if (dlc == NULL || host_config == NULL || uri == NULL) {
 		return -EINVAL;
 	}
 
-	set_state(client, DOWNLOAD_CLIENT_CLOSING);
-	k_sem_give(&client->wait_for_download);
+	k_mutex_lock(&dlc->mutex, K_FOREVER);
 
-	return 0;
-}
+	if (!is_state(dlc, DOWNLOAD_CLIENT_IDLE) &&
+	    !is_state(dlc, DOWNLOAD_CLIENT_CONNECTED)) {
+		LOG_ERR("Invalid start state: %d", dlc->state);
+		k_mutex_unlock(&dlc->mutex);
+		return -EPERM;
+	}
 
-int download_client_start(struct download_client *client, const char *file,
-			  size_t from)
-{
-	if (client == NULL) {
-		return -EINVAL;
+	/* Check if we are already connected to the correct host */
+	if (is_state(dlc, DOWNLOAD_CLIENT_CONNECTED)) {
+		char hostname[CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE];
+
+		err = url_parse_host(uri, hostname, sizeof(hostname));
+		if (err) {
+			LOG_ERR("Failed to parse hostname");
+			return EINVAL;
+		}
+		if (strncmp(hostname, dlc->hostname, sizeof(hostname)) == 0) {
+			transport_connected = dlc->transport;
+		}
 	}
 
-	k_mutex_lock(&client->mutex, K_FOREVER);
+	/* Extract the hostname, without protocol or port */
+	err = url_parse_host(uri, dlc->hostname, sizeof(dlc->hostname));
+	if (err) {
+		LOG_ERR("Failed to parse hostname");
+		return EINVAL;
+	}
 
-	if (client->host == NULL) {
-		k_mutex_unlock(&client->mutex);
-		return -EINVAL;
+	/* Extract the filename, without protocol or port */
+	err = url_parse_file(uri, dlc->file, sizeof(dlc->file));
+	if (err) {
+		LOG_ERR("Failed to parse filename");
+		return EINVAL;
 	}
 
-	if (!is_idle(client) && !is_finished(client)) {
-		k_mutex_unlock(&client->mutex);
-		return -EALREADY;
+	dlc->host_config = *host_config;
+	dlc->file_size = 0;
+	dlc->progress = from;
+	dlc->buf_offset = 0;
+
+	dlc->transport = NULL;
+
+	STRUCT_SECTION_FOREACH(dlc_transport_entry, entry) {
+		if (entry->transport->proto_supported(dlc, uri)) {
+			dlc->transport = entry->transport;
+			break;
+		}
 	}
 
-	client->file = file;
-	client->file_size = 0;
-	client->progress = from;
-	client->offset = 0;
-	client->http.has_header = false;
-	if (is_idle(client)) {
-		set_state(client, DOWNLOAD_CLIENT_CONNECTING);
+	if (!dlc->transport) {
+		LOG_WRN("Protocol not specified for %s, attempting http", uri);
+		STRUCT_SECTION_FOREACH(dlc_transport_entry, entry) {
+			if (entry->transport->proto_supported(dlc, "http://")) {
+				dlc->transport = entry->transport;
+				break;
+			}
+		}
+
+		if (!dlc->transport) {
+			LOG_ERR("Protocol not specified for %s and http could not be used", uri);
```

Comment:
```suggestion
			LOG_ERR("Protocol not found for %s", uri);
```

As `coap://example.com/file` might fail if CoAP transport is off.  Previous line already told that `http://` is tried.